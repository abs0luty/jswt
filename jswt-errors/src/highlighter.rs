use colored::{Color, Colorize};
use regex::Regex;

/// Highlighter that functions as a mini tokenizer.
/// Rather than return a token stream just apply the highlighting
/// in place.
struct HighlightRule {
    matcher: Regex,
    color: Color,
}

pub fn highlight(source: &str) -> String {
    let rules = [
        HighlightRule {
            // Keywords
            matcher: Regex::new(r"^(\bfunction|let|const|return)\b").unwrap(),
            color: Color::Cyan,
        },
        HighlightRule {
            // built-ins
            matcher: Regex::new(r"^\b(i32|f32|string)\b").unwrap(),
            color: Color::Yellow,
        },
        HighlightRule {
            // Ident
            matcher: Regex::new(r"^\b[a-z_][A-Za-z0-9]*\b").unwrap(),
            color: Color::White,
        },
        HighlightRule {
            // Class?
            matcher: Regex::new(r"^\b[A-Za-z_][A-Za-z0-9]*\b").unwrap(),
            color: Color::Yellow,
        },
        HighlightRule {
            matcher: Regex::new(r"^\d+").unwrap(),
            color: Color::Magenta,
        },
    ];

    let mut offset = 0;

    let mut highlighted_source = String::from("");
    'outer: while offset < source.len() {
        let rest = &source[offset..];

        for rule in rules.iter() {
            if let Some(res) = rule.matcher.find(rest) {
                let match_text = res.as_str();
                // Advance cursor based on match
                offset += match_text.len();
                highlighted_source += &match_text.color(rule.color).to_string();
                continue 'outer;
            }
        }

        highlighted_source += &source[offset..offset + 1].clear().to_string();
        offset += 1;
    }

    highlighted_source.to_owned()
}
