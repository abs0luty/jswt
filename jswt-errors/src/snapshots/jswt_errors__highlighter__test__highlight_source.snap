---
source: jswt-errors/src/highlighter.rs
expression: actual

---

        [90m// @ts-nocheck[0m

        [90m// Single max utility function[0m
        [36mfunction[0m [37mmax[0m([37ma[0m: [33mi32[0m, [37mb[0m: [33mi32[0m): [33mi32[0m {
            [37mif[0m ([37ma[0m > [37mb[0m) {
                [36mreturn[0m [37ma[0m;
            }
            [36mreturn[0m [37mb[0m;
        }
        
        [36mfunction[0m [37mmin[0m([37ma[0m: [33mi32[0m, [37mb[0m: [33mi32[0m): [33mi32[0m {
            [37mif[0m ([37ma[0m < [37mb[0m) {
                [36mreturn[0m [37ma[0m;
            }
            [36mreturn[0m [37mb[0m;
        }
        
        @[37minline[0m
        @[37mwast[0m([32m"(i32.lt_u (local.get $a) (local.get $b))"[0m)
        [36mfunction[0m [37mlessUnsigned[0m([37ma[0m: [33mi32[0m, [37mb[0m: [33mi32[0m): [37mbool[0m { }
        
        @[37minline[0m
        @[37mwast[0m([32m"(i32.ge_u (local.get $a) (local.get $b))"[0m)
        [36mfunction[0m [37mgreaterEqUnsigned[0m([37ma[0m: [33mi32[0m, [37mb[0m: [33mi32[0m): [37mbool[0m { }
        
        @[37minline[0m
        @[37mwast[0m([32m"(i32.gt_u (local.get $a) (local.get $b))"[0m)
        [36mfunction[0m [37mgreaterUnsigned[0m([37ma[0m: [33mi32[0m, [37mb[0m: [33mi32[0m): [37mbool[0m { }
        
        @[37minline[0m
        @[37mwast[0m([32m"(i32.div_u (local.get $a) (local.get $b))"[0m)
        [36mfunction[0m [37mdivUnsigned[0m([37ma[0m: [33mi32[0m, [37mb[0m: [33mi32[0m): [37mbool[0m { }
        
